### PR TITLE
Added dotAll flag

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-helpers.c [hunk 1]
```diff
@@ -1357,7 +1357,8 @@ void
 ecma_bytecode_deref (ecma_compiled_code_t *bytecode_p) /**< byte code pointer */
 {
   JERRY_ASSERT (bytecode_p->refs > 0);
-  JERRY_ASSERT (!(bytecode_p->status_flags & CBC_CODE_FLAGS_STATIC_FUNCTION));
+  JERRY_ASSERT (!CBC_IS_FUNCTION (bytecode_p->status_flags) || \
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers.c:1360]: `||` should be in the new line, and no `\` is needed.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 2]
```diff
@@ -1450,8 +1457,11 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
 
         const lit_code_point_t cp = ecma_regexp_unicode_advance (&str_curr_p, re_ctx_p->input_end_p);
 
-        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX && lit_char_is_line_terminator ((ecma_char_t) cp)))
+        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX &&
+            lit_char_is_line_terminator ((ecma_char_t) cp) &&
+            !(re_ctx_p->flags & RE_FLAG_DOTALL)))
         {
+
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1464]: No new-line, here.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 3]
```diff
@@ -1450,8 +1457,11 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
 
         const lit_code_point_t cp = ecma_regexp_unicode_advance (&str_curr_p, re_ctx_p->input_end_p);
 
-        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX && lit_char_is_line_terminator ((ecma_char_t) cp)))
+        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX &&
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1460]: if `(re_ctx_p->flags & RE_FLAG_DOTALL)` present, the remaining conditions should not be evaluated.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 4]
```diff
@@ -1466,8 +1476,11 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
         }
 
         const ecma_char_t ch = lit_cesu8_read_next (&str_curr_p);
-
-        if (lit_char_is_line_terminator (ch))
+        bool has_dot_all_flag = false;
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1479]: This should be es5.1 guarded.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 5]
```diff
@@ -1466,8 +1476,11 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
...
+#if ENABLED (JERRY_ESNEXT)
+        has_dot_all_flag = re_ctx_p->flags & RE_FLAG_DOTALL;
+#endif /* ENABLED (JERRY_ESNEXT) */
+        if (lit_char_is_line_terminator (ch) && !has_dot_all_flag)
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1483]: swap the conditions order.

File: jerry-core/parser/js/js-lexer.c [hunk 6]
```diff
@@ -2799,7 +2799,12 @@ lexer_construct_regexp_object (parser_context_t *context_p, /**< context */
     {
       flag = RE_FLAG_STICKY;
     }
-
```
Comment (rerobika) [inline jerry-core/parser/js/js-lexer.c:2802]: do not remove this line.

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.inc.h [hunk 7]
```diff
@@ -80,6 +84,11 @@ SIMPLE_VALUE (LIT_MAGIC_STRING_MULTILINE,
               ECMA_VALUE_FALSE,
               ECMA_PROPERTY_FIXED)
 
+/* ECMA-262 v9, 21.2.5.3 */
+SIMPLE_VALUE (LIT_MAGIC_STRING_DOTALL,
```
Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.inc.h:88]: This is 5.1 code, this should not be added here.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 8]
```diff
@@ -1450,11 +1457,15 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
...
+        if (re_ctx_p->flags & RE_FLAG_DOTALL)
+        {
+          continue;
+        }
+        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX &&
+            lit_char_is_line_terminator ((ecma_char_t) cp)))
         {
           goto fail;
         }
-
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1457]: Do not remove this line.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 9]
```diff
@@ -1466,8 +1477,12 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
         }
 
         const ecma_char_t ch = lit_cesu8_read_next (&str_curr_p);
-
-        if (lit_char_is_line_terminator (ch))
+#if !ENABLED (JERRY_ESNEXT)
+        bool has_dot_all_flag = false;
+#else
+        bool has_dot_all_flag = re_ctx_p->flags & RE_FLAG_DOTALL;
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1483]: (re_ctx_p->flags & RE_FLAG_DOTALL) != 0;

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 10]
```diff
@@ -1466,8 +1477,12 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
         }
 
         const ecma_char_t ch = lit_cesu8_read_next (&str_curr_p);
-
-        if (lit_char_is_line_terminator (ch))
+#if !ENABLED (JERRY_ESNEXT)
+        bool has_dot_all_flag = false;
+#else
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1482]: ` /* ENABLED (JERRY_ESNEXT) */`

**Comment by rerobika** [comment]
Please rebase after #3979.

**Comment by kisbg** [comment]
@dbatyai  @rerobika  thank you for the review! i updated  the pull request

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 11]
```diff
@@ -1505,11 +1512,11 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
 
         const lit_code_point_t cp = ecma_regexp_unicode_advance (&str_curr_p, re_ctx_p->input_end_p);
 
-        if (JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX && lit_char_is_line_terminator ((ecma_char_t) cp)))
+        if (JERRY_UNLIKELY (!(re_ctx_p->flags & RE_FLAG_DOTALL) && cp <= LIT_UTF16_CODE_UNIT_MAX &&
+            lit_char_is_line_terminator ((ecma_char_t) cp)))
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1516]:
```c
if (!(re_ctx_p->flags & RE_FLAG_DOTALL)
    && JERRY_UNLIKELY (cp <= LIT_UTF16_CODE_UNIT_MAX 
                       && lit_char_is_line_terminator ((ecma_char_t) cp)))
```

File: jerry-core/ecma/operations/ecma-regexp-object.c (hunk 8 above)
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-regexp-object.c:1512]: This request is still valid.

File: jerry-core/ecma/operations/ecma-regexp-object.c [hunk 12]
```diff
@@ -1521,8 +1528,12 @@ ecma_regexp_run (ecma_regexp_ctx_t *re_ctx_p, /**< RegExp matcher context */
...
 
         const ecma_char_t ch = lit_cesu8_read_next (&str_curr_p);
-
-        if (lit_char_is_line_terminator (ch))
+#if !ENABLED (JERRY_ESNEXT)
+        bool has_dot_all_flag = false;
+#else /* ENABLED (JERRY_ESNEXT) */
+        bool has_dot_all_flag = (re_ctx_p->flags & RE_FLAG_DOTALL) != 0;
+#endif /* !ENABLED (JERRY_ESNEXT) */
+        if (!has_dot_all_flag && lit_char_is_line_terminator (ch))
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-regexp-object.c:1536]: Newline before.

**Pull request merged**